lib/event_store_client/connection.rb

```diff
@@ -79,5 +70,51 @@ def client
           host: host, port: port, per_page: per_page
         )
     end
+
+    def read_from_stream(stream, direction:, start:, resolve_links:)
+      response =
+        client.read(
+          stream, start: start, direction: direction, resolve_links: resolve_links
+        )
+      return [] if response.body.nil? || response.body.empty?
+      JSON.parse(response.body)['entries'].map do |entry|
+        deserialize_event(entry)
+      end.reverse
+    end
+
+    def read_all_from_stream(stream, start:, resolve_links:)
+      count = per_page
+      events = []
+      failed_requests_count = 0
+
+      while failed_requests_count < 3
```

Comment:
I would split while part into methods
I would prepare fetch_data(or similar name) where I would pass needed params. I would put while loop there with handling errors from Faraday and failed requests and return a response.